crates/transaction-pool/src/pool/parked.rs

```diff
@@ -79,10 +87,98 @@ impl<T: ParkedOrd> ParkedPool<T> {
         Some(tx.transaction.into())
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.best.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Get transactions by sender
+    pub(crate) fn get_txs_by_sender(
+        &self,
+        sender: &Address,
+    ) -> Vec<Arc<ValidPoolTransaction<T::Transaction>>> {
+        self.by_id
+            .values()
+            .filter(|tx| tx.transaction.sender() == *sender)
+            .map(|tx| tx.transaction.clone().into())
+            .collect()
+    }
+
+    /// Returns addresses sorted by their last submission id. Addresses with older last submission
+    /// ids are first. Note that _last_ submission ids are the newest submission id for that
+    /// sender, so this sorts senders by the last time they submitted a transaction in descending
+    /// order. Senders that have least recently submitted a transaction are first.
+    ///
+    /// Similar to `Heartbeat` in Geth
+    pub(crate) fn get_senders_by_submission_id(&self) -> Vec<Address> {
+        // iterate through by_id, and get the last submission id for each sender
+        let senders = self
+            .by_id
+            .iter()
+            .fold(Vec::new(), |mut set: Vec<Reverse<(u64, Address)>>, (_, tx)| {
+                if let Some(last) = set.last_mut() {
+                    // sort by last
+                    if last.0 .1 == tx.transaction.sender() {
```

Comment:
this should use `SenderId` not address, since internally we only really care about the Id

Reply:
good idea, now this uses `SenderId`